File: FinLib.NET/FinLib/Price/PricePair.cs
using System;

namespace FinLib.Price;

public class PricePair: IDisposable
{
    private readonly unsafe PricePair_native* _price;
    internal unsafe PricePair_native* GetPtr() => _price;

    internal unsafe PricePair(PricePair_native* price)
    {
        _price = price;
    }

    public PricePair(float bid, float offer)
    {
        unsafe
        {
            _price = NativeMethods.price_pair_new(bid, offer);
        }
    }

    public double Bid
    {
        get
        {
            unsafe
            {
                return NativeMethods.price_pair_get_bid(_price);
            }
        }
        set
        {
            unsafe
            {
                NativeMethods.price_pair_set_bid(_price, value);
            }
        }
    }

    public double Offer
    {
        get
        {
            unsafe
            {
                return NativeMethods.price_pair_get_offer(_price);
            }
        }
        set
        {
            unsafe
            {
                NativeMethods.price_pair_set_offer(_price, value);
            }
        }
    }

    public double Spread {
        get
        {
            unsafe
            {
                return NativeMethods.price_pair_spread(_price);
            }
        }
    }

    public double Midpoint {
        get
        {
            unsafe
            {
                return NativeMethods.price_pair_midpoint(_price);
            }
        }
    }

    private void ReleaseUnmanagedResources()
    {
        unsafe
        {
            NativeMethods.price_pair_destroy(_price);
        }
    }

    public void Dispose()
    {
        ReleaseUnmanagedResources();
        GC.SuppressFinalize(this);
    }

    ~PricePair()
    {
        ReleaseUnmanagedResources();
    }
}

File: finlib-cpp/src/finlib/price/price_pair.cpp
//
// Created by Andy Pack on 19/06/2025.
//

#include "price_pair.hpp"

#include "finlib/finlib-native.h"

namespace finlib {
    PricePair::PricePair(double bid, double offer) {
        handle = finlibrs::price_pair_new(bid, offer);
    }
    PricePair::PricePair(finlibrs::PricePair *handle) {
        this->handle = handle;
    }
    double PricePair::spread() {
        return finlibrs::price_pair_spread(handle);
    }
    double PricePair::midpoint() {
        return finlibrs::price_pair_midpoint(handle);
    }

    PricePair::~PricePair() {
        finlibrs::price_pair_destroy(handle);
    }
} // finlib